Reorganize

dalehamel opened the pull request.

dalehamel committed:
Add vernier backend support

dalehamel committed:
Fixup

dalehamel commented on Gemfile:
```diff
@@ -6,6 +6,7 @@ gemspec
 # Specify the same dependency sources as the application Gemfile
 gem("activesupport", "~> 5.2")
 gem("railties", "~> 5.2")
+gem("vernier", "~> 0.3.1") if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("3.2.1")
```
NIT bump to 0.4.0

dalehamel commented on README.md:
```diff
@@ -312,6 +314,22 @@ Note that in `development` and `test` modes the file isn't uploaded. Instead, it
+
+By default, the stackprof backend will be used.
+
+In local development, changing the backend will change whether the profile is viewed in speedscope, or firefox-profiler.
```
NIT move this to viewer PR

bmansoob replied:
```suggestion
In local development, changing the backend will change whether the profile is viewed in speedscope or firefox-profiler.
```

dalehamel commented on lib/app_profiler/backend.rb:
```diff
@@ -0,0 +1,33 @@
+  end
+end
+
+require "app_profiler/backend/stackprof"
```
TODO see if these can be autoloaded and moved to top of module

dalehamel commented on lib/app_profiler/backend/vernier.rb:
```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+begin
+  gem("vernier", ">= 0.3.1")
```
NIT bump to 0.4.0

dalehamel commented on test/app_profiler/profile/vernier_test.rb:
```diff
@@ -0,0 +1,153 @@
+      assert_equal("mock", profile.id)
+    end
+
+    # test ".from_vernier removes id and context metadata from profile data" do
```
NIT remove this and other commented tests

bmansoob commented on README.md:
```diff
@@ -76,12 +76,14 @@ Rails.application.config.app_profiler.profile_header = "X-Profile"
 | ignore_gc | Ignore garbage collection frames | |
 | autoredirect | Redirect request automatically to Speedscope's page after profiling. | |
 | context | Directory within the specified bucket in the selected storage where raw profile data should be written. | Only supported in (2). Defaults to `Rails.env` if not specified. |
+| backend | Profiler to use, either `stackprof` or `vernier`. Defaults to stackprof. Available only vernier is installed in the app, and Ruby version is at least 3.2.1 |
```
```suggestion
| backend | Profiler to use, either `stackprof` or `vernier`. Defaults to stackprof. Vernier is only available when added separately in the application Gemfile, and the Ruby version is at least 3.2.1 |
```

bmansoob commented on README.md:
```diff
@@ -312,6 +314,22 @@ Note that in `development` and `test` modes the file isn't uploaded. Instead, it
+
+It is possible to configure app_profiler to use the [`vernier`](https://github.com/jhawthorn/vernier) or stackprof. To use vernier, it must be added separately in the application Gemfile.
+
+The backend can be selected at dynamicall runtime using the `backend` parameter. The default backend to use when this parameter is not specified can be configured with:
```
```suggestion
The backend can be selected dynamically at runtime using the `backend` parameter. The default backend to use when this parameter is not specified can be configured with:
```

bmansoob commented on lib/app_profiler/backend.rb:
```diff
@@ -0,0 +1,33 @@
+      raise NotImplementedError
+    end
+
+    def start(params = {})
```
I think thins carrying over from before, but I wonder why do we have both `start` and `run` as public? Looks like `run` is the one used by the middleware. 

You do not have to change it in this PR, just curious.

dalehamel replied:
I think this is in case someone ever wants to basically just directly profile something using a similar API to stackprof, ie, explicitly calling start and stop rather than using a block. And yeah, it is from before.

bmansoob commented on lib/app_profiler/backend/stackprof.rb:
```diff
@@ -0,0 +1,89 @@
+      # Do not start the profiler if StackProf was started somewhere else.
+      return false if running?
+
+      clear
```
maybe we should `clear` in the ensure block up in `run`?

bmansoob commented on lib/app_profiler/backend.rb:
```diff
@@ -0,0 +1,33 @@
+      raise NotImplementedError
+    end
+
+    def stop
```
another one which I am 🤔 -- looks like we will always use stop followed by results.

bmansoob commented on lib/app_profiler/backend/stackprof.rb:
```diff
@@ -0,0 +1,89 @@
+      nil
+    end
+
+    def running?
```
wonder if we need to treat this as a critical section

bmansoob commented on lib/app_profiler/middleware.rb:
```diff
@@ -24,13 +25,31 @@ def call(env, params = AppProfiler::RequestParameters.new(Rack::Request.new(env)
+      @backend_lock.synchronize do
+        if params.backend &&
+            AppProfiler.profiler_backend.name.split("::").last.downcase.gsub("backend", "") != params.backend
+          raise ArgumentError if AppProfiler.running?
```
may move this up and change to `yield if AppProfiler.running?`

bmansoob commented on lib/app_profiler/middleware.rb:
```diff
@@ -24,13 +25,31 @@ def call(env, params = AppProfiler::RequestParameters.new(Rack::Request.new(env)
       return yield unless before_profile(env, params_hash)
 
+      @backend_lock.synchronize do
+        if params.backend &&
```
I wonder why do we need this - if somehow we end up calling`profile` twice?

dalehamel replied:
My worry is Puma, if we have one thread that is being profiled by Stackprof, we cannot allow it to switch to Vernier while stackprof is running for instance.

bmansoob commented on lib/app_profiler/middleware.rb:
```diff
@@ -24,13 +25,31 @@ def call(env, params = AppProfiler::RequestParameters.new(Rack::Request.new(env)
+          raise ArgumentError if AppProfiler.running?
+
+          orig_backend = AppProfiler.profiler_backend
+          if defined?(AppProfiler::VernierBackend) &&
```
Can we ✂️ `AppProfiler::*` and just use `VernierBackend` etc?

bmansoob commented on lib/app_profiler/middleware.rb:
```diff
@@ -24,13 +25,31 @@ def call(env, params = AppProfiler::RequestParameters.new(Rack::Request.new(env)
 
+      @backend_lock.synchronize do
+        if params.backend &&
+            AppProfiler.profiler_backend.name.split("::").last.downcase.gsub("backend", "") != params.backend
```
I think this validation should move here
https://github.com/Shopify/app_profiler/blob/9eb54130da5adf6f731f2c5aee86fcfebc1e7915/lib/app_profiler/server.rb#L127

dalehamel replied:
Maybe in request_params instead? The server code is only for the standalone server

bmansoob replied:
https://github.com/Shopify/app_profiler/blob/9eb54130da5adf6f731f2c5aee86fcfebc1e7915/lib/app_profiler/request_parameters.rb#L19 yeah here, sorry I linked the wrong thing.

dalehamel committed:
Apply suggestions from code review

Co-authored-by: Bassam Mansoob <bassam.mansoob@shopify.com>

dalehamel committed:
Move backend updating code into main AppProfiler module

bmansoob commented on lib/app_profiler/middleware.rb:
```diff
@@ -24,13 +24,26 @@ def call(env, params = AppProfiler::RequestParameters.new(Rack::Request.new(env)
+        orig_backend = AppProfiler.profiler_backend
+        if defined?(AppProfiler::VernierBackend) &&
+            params.backend == AppProfiler::VernierBackend::NAME
+          return yield unless (AppProfiler.backend = AppProfiler::VernierBackend)
```
`return yield unless (AppProfiler.backend = AppProfiler::VernierBackend)` I am not sure I understand why do we need this 🤔

bmansoob replied:
is it because `AppProfiler#backend=` can raise if `running?` but that is not rescued here.

I am starting to think if we are better of by doing this `running?` check in the middleware, and just one place. Set some car to true before running the profiling code and reset it before sending the response. For a multithreaded app, if `running?` is true for another request, we just do not execute the profiling code.

This assumes that no code will interact with backends directly, which I think they should not.

dalehamel replied:
yeah i think actually it would be better to just return "false" from `backend=` I realize now that raising without an ensure block is bad. The intention was that if we were not able to change the backend, but that was requested, to just run the code and fail to profile.

I agree that it's getting a bit hairy with the `running?` checks, because that is usually where the edge cases tend to be.

bmansoob commented on lib/app_profiler/backend/stackprof.rb:
```diff
@@ -0,0 +1,89 @@
+
+      yield
+
+      return unless started
```
can simplify this is a bit, I think.

`yield unless started `

dalehamel replied:
FYI this file was copied over verbatim, was formerly "profiler.rb".

dalehamel committed:
Refactor access to setting backend

profiler_backend is now private, and can only be read and updated through:

- AppProfiler.backend=
- AppProfiler.backend

A new helper method AppProfiler.with_backend is used to wrap singular access
to the backend.

dalehamel committed:
Fix vernier load error

dalehamel committed:
Raise when viewing vernier

dalehamel committed:
Readme fix

dalehamel committed:
Remove commented code

bmansoob commented on lib/app_profiler.rb:
```diff
@@ -79,25 +78,41 @@ def running?
+      if backend
+        if defined?(AppProfiler::VernierBackend) &&
+            backend == AppProfiler::VernierBackend::NAME
+          return yield unless (AppProfiler.backend = AppProfiler::VernierBackend)
```
`backend=` raises when it is `running?` I think we need to either return `nil` or rescue, I prefer the former. 

Also the way `with_backend` is used in the middleware:

```
 AppProfiler.with_backend(params.backend) do
        AppProfiler.run(params_hash) do
         ......
```

...`return yield unless...` means that we will end up running `AppProfiler.run`? I think with_backend can just return `nil` or the actual backend, and then we can handle it in the middleware.

dalehamel replied:
> backend= raises when it is running? I think we need to either return nil or rescue, I prefer the former.

Good call, i missed this

> I think with_backend can just return nil or the actual backend, and then we can handle it in the middleware.

I'm a bit confused by this, how would it help to expose the backend back to the calling layer? And in the middleware, how would we get access to the profile when calling `AppProfiller.run` inside the `with_backend` block? Normally i think when wrapping using a `with_xxx` you would just return the value of the block 🤔

dalehamel committed:
Pull with_backend code into AppProfiler.run

dalehamel commented on lib/app_profiler.rb:
```diff
@@ -60,17 +63,73 @@ module Viewer
+    end
+
+    def backend=(new_backend)
+      new_profiler_backend = if new_backend.is_a?(String)
```
a bit weird i know, but when we have the middleware params we only have a string. I'd rather do the lookup here.

It also feels silly to require that other calls send the string via ::NAME, so i thought may as well take advantage of ruby duck typing here and let it accept either, and resolve it to the type it needs.

dalehamel committed:
Lock around AppProfiler.run
